graph_analysis/a_b_charts.py

```diff
@@ -0,0 +1,222 @@
+        scatter_plot = alt.Chart(point_df[point_df.path < max_paths].reset_index()).mark_point(
+            size=400,
+            opacity=1,
+            filled=True
+        ).encode(
+            x=alt.X('x:Q', sort=None, axis=None),
+            y=alt.Y('y:Q', sort=None, axis=None),
+            color=alt.Color("path", scale=alt.Scale(domain=list(range(max_paths)), range=self.colors[:max_paths]), legend=None),
+            opacity=alt.condition(point_selection, alt.value(1), alt.value(0.4)),
+            shape=alt.Shape('type:N', scale=alt.Scale(range=['circle', 'diamond']), legend=None),
+            tooltip="index"
+            ).add_selection(point_selection)
+        return scatter_plot
+
+    def get_graph_chart(self, max_paths=3):
+        """Combines line charts with a scatter plot and returns full graph chart."""
+        line_charts = [self.draw_path(self.paths_df[i], self.colors[i]) for i in range(min(max_paths,
+                                                                                  len(self.paths_df)))][::-1]
+        scatter_plot = self.draw_nodes(self.point_df)
+        base_chart = line_charts[0]
+        for chart in line_charts[1:]:
+            base_chart += chart
+        base_chart += scatter_plot
+        return base_chart.properties(
+            width=900,
+            background = "#f5f3f2",
+            padding=30,
+            title=alt.TitleParams(text="All the paths between dataset A and dataset B.",
+                                  subtitle=["Shortest path is displayed in the middle."],
+                                  font="Helvetica Neue",
+                                  fontSize=22,
+                                  color="#3e454f",
+                                  subtitleFont="Helvetica Neue",
+                                  subtitleFontSize=16,
+                                  subtitleColor="#3e454f",
+                                  anchor="start",
+                                  dx=20,
+                                  dy=-20
+                                  )
+            ).configure_axis(
+                ticks=False,
+                labelFont="Helvetica Neue",
+                labelFontSize=14,
+                labelColor="#494F59",
+                labelPadding=10,
+                title=None
+            ).configure_view(strokeOpacity=0)
+
+    def get_slo_chart(self):
+        """Creates a multiline chart with points for slo visualization."""
+        total_slo = list(self.slo_df.slo)[-1]
+        if total_slo < 6 * 60:
+            label_interval, label_variable, time_div = 10, "s", 1
+        elif total_slo < 100 * 60:
+            label_interval, label_variable, time_div = 60, "m", 60
+        elif total_slo < 100 * 60 * 60:
+            label_interval, label_variable, time_div = 60 * 60, "h", 60 * 60
+        else:
+            label_interval, label_variable, time_div =  24 * 60 * 60, "d",  24 * 60 * 60
```

Comment:
These are manually configured values, to start visualising minutes if having more than 360 seconds, hours after 100 minutes, and days otherwise.
Should be 10, but can be changed, will add more comments explaining.